solutions/usaco-718.mdx

```diff
@@ -0,0 +1,92 @@
+---
+id: usaco-718
+source: USACO Gold 2017 February
+title: Why Did the Cow Cross The Road II
+author: Brad Ma
+---
+ 
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_nocross_gold_feb17.html)
+
+The recursive approach where you try all configurations of crosswalks has the 
+time complexity $\mathcal{O}(2 ^ N)$, which is extremely slow. We observe that we have 
+many duplicate recursive calls where the configuration has been solved already, so by 
+using a top down dynamic progamming approach with memoization we can significantly reduce 
+the time complexity to $\mathcal{O}(N ^ 2)$. In this approach, we store our answers such 
+that if there ever is a call to it again, the function uses the cached answer instead of 
+recomputing it. A bottom up dynamic progamming approach without recursion can run faster 
+because function calls have a large constant, but it is more difficult to implement and 
+is not necessary to pass all test cases.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N ^ 2)$
+
+<LanguageSection>
+  
+<JavaSection>
+  
+```java
+import java.io.*;
+import java.util.*;
+
+public class WhyDidTheCowCrossTheRoadII {
+	static int[] firstSide;
+	static int[] secondSide;
+	static int[][] memoization;
+
+	static boolean friendly (int n, int m) {
+		if (Math.abs(firstSide[n] - secondSide[m]) <= 4) {
+			return true;
+		}
+		return false;
```

Comment:
just `return Math.abs(firstSide[n] - secondSide[n]) <= 4;`

Reply:
that will make it less readable. it isn't always about making it the shortest lines